src/Swarm/Game/Challenge.hs

```diff
@@ -146,5 +150,15 @@ mkWorldFun pwd = E $ \em -> do
       . mapM toEntity
       . concat
       $ grid
-  let defTerrain = (fromEnum *** (>>= (`lookupEntityName` em))) (defaultTerrain wd)
-  return $ worldFunFromArray arr defTerrain
+  case defaultTerrain wd of
+    Left seed -> return $ \entities ->
+      let emPlus = em <> entities
+          arr2 = bimap toEnum (fmap (^. entityName)) <$> arr
+       in (lkup emPlus <$>) . first fromEnum <$> testWorld2FromArray arr2 seed
+    Right def ->
+      let defTerrain = (fromEnum *** (>>= (`lookupEntityName` em))) def
+       in return . const $ worldFunFromArray arr defTerrain
```

Comment:
Actually it might be a good idea to unify this, so that challenges could use the entities which we load anyway. :thinking:

Reply:
I'm not sure I know what you're referring to.  It might be a good idea to unify what?

Reply:
I thought the normal entities are not available at this point.

I probably just got confused by the code 😅

Reply:
I agree it is a little confusing.  I should add some more comments.  Basically, in the `FromJSONE EntityMap Challenge` instance, any custom entities (defined in an `entities` field) are parsed, and those are passed to `mkWorldFun` via `withE` (https://github.com/byorgey/swarm/blob/0623c07f13b108c7b3991fa3b6fd5de31e0592de/src/Swarm/Game/Challenge.hs#L74).  However, `withE` *combines* the given `EntityMap` of custom entities with the ambiently available one (of normal entities) that is passed in when `parseJSONE` is run.  Maybe `appendE` would be a better name?